src/register/pmpcfgx.rs

```diff
@@ -1,23 +1,293 @@
+            index,
+            permission: self.get_permission(index),
+            range: self.get_range(index),
+            locked: self.is_locked(index),
+        }
+    }
+
+    /// PmpByte methods to get a pmp configuration attributes
+    pub fn get_byte(&self, index: usize) -> usize {
+        self.bits.get_bits(8 * index..=(8 * index) + 7) as usize
+    }
+
+    #[inline]
+    pub fn is_locked(&self, index: usize) -> bool {
+        self.bits.get_bit(7 + 8 * index)
+    }
+
+    #[inline]
+    pub fn get_permission(&self, index: usize) -> Option<Permission> {
+        match self.bits.get_bits(8 * index..=8 * index + 2) {
+            0 => Some(Permission::NONE),
+            1 => Some(Permission::R),
+            2 => Some(Permission::W),
+            3 => Some(Permission::RW),
+            4 => Some(Permission::X),
+            5 => Some(Permission::RX),
+            6 => Some(Permission::WX),
+            7 => Some(Permission::RWX),
+            _ => unreachable!(),
+        }
+    }
+
+    #[inline]
+    pub fn get_range(&self, index: usize) -> Option<Range> {
+        match self.bits.get_bits(8 * index + 3..=8 * index + 4) {
+            0 => Some(Range::OFF),
+            1 => Some(Range::TOR),
+            2 => Some(Range::NA4),
+            3 => Some(Range::NAPOT),
+            _ => unreachable!(),
+        }
+    }
+}
+/// Physical memory protection configuration
+/// Pmpcfg0 struct contains pmp0cfg - pmp3cfg for RV32, or pmp0cfg - pmp7cfg for RV64
+/// get_byte() method retrieves a single pmp<x>cfg held in a PmpByte struct
 pub mod pmpcfg0 {
-    read_csr_as_usize!(0x3A0, __read_pmpcfg0);
-    write_csr_as_usize!(0x3A0, __write_pmpcfg0);
+    use super::{Permission, Pmpcfg, Range};
+    use bit_field::BitField;
+
+    read_csr_as!(Pmpcfg, 0x3A0, __read_pmpcfg0);
+    write_csr!(0x3A0, __write_pmpcfg0);
+    set!(0x3A0, __set_pmpcfg0);
+    clear!(0x3A0, __clear_pmpcfg0);
+
+    #[inline]
+    pub unsafe fn set_permissions(permission: Permission, index: usize) {
+        #[cfg(riscv32)]
+        assert!(index < 4);
+
+        #[cfg(riscv64)]
+        assert!(index < 8);
+
+        let mut value = _read();
+        value.set_bits(8 * index..=8 * index + 2, permission as usize);
+        _write(value);
+    }
+
+    #[inline]
+    pub unsafe fn set_range(range: Range, index: usize) {
+        #[cfg(riscv32)]
+        assert!(index < 4);
+
+        #[cfg(riscv64)]
+        assert!(index < 8);
+
+        let mut value = _read();
+        value.set_bits(8 * index + 3..=8 * index + 4, range as usize);
+        _write(value);
+    }
+
+    #[inline]
+    pub unsafe fn set_lock(index: usize) {
+        #[cfg(riscv32)]
+        assert!(index < 4);
+
+        #[cfg(riscv64)]
+        assert!(index < 8);
+
+        _set(1 << (7 + index * 8));
+    }
 }
 
-/// Physical memory protection configuration, RV32 only
+/// Physical memory protection configuration
+/// Pmpcfg1 struct contains pmp4cfg - pmp7cfg for RV32 only
```

Comment:
I wonder if we should limit this by #[cfg] for RV64 so it's completely inaccessible on that target.

Reply:
I was just thinking about this last night.  Totally agree.  I will make the change.